Add ytt getting started and leigh (#1198)
vmware-tanzu/tanzu-dev-portal: content/guides/kubernetes/ytt-gs.md

---
date: '2021-08-16'
lastmod: '2021-08-16'
parent: Carvel
patterns:
- Deployment

> tags:

Reviewer: add `description` to front matter, please :)

- ytt
- Carvel
team:
- Leigh Capili
- Tiffany Jernigan
title: Getting Started with ytt
topics:
- Kubernetes
oldPath: "/content/guides/kubernetes/ytt-gs.md"
aliases:
- "/guides/kubernetes/ytt-gs"
level1: Deploying Modern Applications
level2: CI/CD, Release Pipelines
---

[ytt](https://carvel.dev/ytt/) (part of the open source [Carvel](https://carvel.dev) suite) is a tool that can templatize, combine, and patch any YAML content.
ytt understands the structure of YAML. 
This eliminates the need to count spaces or escape quotes in traditional text-based templating.
ytt uses a sandboxed, Pythonic language, embedded within YAML comments, to provide a natural, programming extension for YAML.

ytt can simplify and variate any YAML content, including Kubernetes configuration, Concourse pipelines, Docker Compose files, and more.

This guide will cover the basics of using ytt to configure Kubernetes apps.
## Prerequisites
Before you get started you will need to do the following:
* Create a Kubernetes cluster

> * Install `kubectl` locally

Reviewer 2: ```suggestion
* Install [`kubectl`](https://kubernetes.io/docs/tasks/tools/) locally
```

* Install the `ytt` CLI via one of these options:

> * [Homebrew](https://github.com/vmware-tanzu/homebrew-carvel): 

Reviewer 2: ```suggestion
   * [Homebrew](https://github.com/vmware-tanzu/homebrew-carvel): 
```

    ```
    brew tap vmware-tanzu/carvel
    brew install ytt
    ```

> 	* [GitHub releases](https://github.com/vmware-tanzu/carvel-ytt/releases/tag/v0.36.0): move it to `/usr/local/bin` or add it to your `$PATH` and run `chmod +x` to make it executable

Reviewer 2: ```suggestion
   * [GitHub releases](https://github.com/vmware-tanzu/carvel-ytt/releases/tag/v0.36.0): move it to `/usr/local/bin` or add it to your `$PATH` and run `chmod +x` to make it executable
```

Reviewer: From the VMware style guide, lists formatted like this need to be separated by a "--" as opposed to a ":"

Author: That is extremely specific...wow.

> Note: This guide uses ytt v0.36.0, and we suggest you download and install the same version to ensure the best experience as you follow through this guide.

## Clone the Examples Repo
Now that we have ytt installed, let’s clone the main ytt repo so we can look at some examples:
```shell
git clone https://github.com/vmware-tanzu/carvel-ytt
cd carvel-ytt/examples
```

## Usage
You can execute ytt by running the command on individual files or a directory of files.
ytt will build the configuration and write it to STDOUT by default.
Run the following command to see an example with a plain YAML file.

```shell
ytt -f playground/basics/example-plain-yaml
```

We can use the interactive playground for this YAML at the following link. It will show all files for the example and you can then run them and see the output you’d see with running the ytt CLI. Going forward you can choose to use the ytt CLI or the playground for the majority of the examples.

> https://carvel.dev/ytt/#example:example-plain-yaml 

Reviewer 2: ```suggestion
You can see this file in the playground here: https://carvel.dev/ytt/#example:example-plain-yaml 
```

One example of using an if statement is as follows:
```
#@ if True:
test1: 123
#@ else:
test2: 124
#@ end
```

And here is a for loop:
```
array:
#@ for i in range(0,3):
- #@ i
- #@ i+1
#@ end
```
## Variables, Functions, and Fragments
In this example variables are used to store the app name and version in a single place.
A function is then used to define a repeatable set of labels using a [YAML Fragment](https://carvel.dev/ytt/docs/latest/lang-ref-yaml-fragment/) with an optional parameter to include the version number.

Variable definition:
```yaml
#@ app_name = "prometheus-operator"
#@ version = "v0.39.0"
```

Function with YAML Fragment in it:
```yaml
#@ def labels(with_version=False):
  app.kubernetes.io/component: controller
  app.kubernetes.io/name: #@ app_name
  #@ if with_version:
  app.kubernetes.io/version: #@ version
  #@ end
#@ end
```

ytt comments are then used to calculate Starlark expressions to insert and concatenate this name, label, and version information into YAML nodes in multiple places like the metadata, labelSelector, and container image tag.

```yaml

> #@version

Reviewer 3: this line should have a space: `#@ version`

#@ labels(with_version=True)
#@ app_name + "-service"
```

> Look at file and run in ytt playground:
> https://carvel.dev/ytt/#example:example-extract-yaml-fragments 

Reviewer 2: ```suggestion
[Look at this example](https://carvel.dev/ytt/#example:example-extract-yaml-fragments) and run it in the ytt playground:
```

Author: Does it make it less obvious when done like that that that link is for the playground and is showing the files? And that it's one of the two options?

Reviewer 2: Yeah I debated with myself back and forth on it for a bit, I'm happy to keep it as is too

> Or run locally:

Reviewer 2: ```suggestion
Or run it locally:
```

```shell
ytt -f playground/getting-started/example-extract-yaml-fragments/config.yml
```

## Load
We saw `#@ load` when we were first looking at directives.
So what does `load` do? 
Basically it is used to load functions from other libraries. These could be [built-in modules](https://carvel.dev/ytt/docs/latest/lang-ref-ytt/) or ones you define in another file.
```
#@ load("@ytt:template", "template")
#@ load("@ytt:data", "data")

#@ load("helpers.lib.yml", "mod")

```

## Overlays
ytt has a feature and module called [overlay](https://carvel.dev/ytt/docs/latest/lang-ref-ytt-overlay/#overview) which allows you to patch YAML structures. 
By patching, you could use it for matching, inserting, replacing or removing items, etc. 
They are used to update YAML nodes which don’t have any externally exposed ways to configure data.
Overlays are applied in the order provided. 

> To use an overlay, you need to load the built-in overlay function.

Reviewer 2: ```suggestion
To use an overlay, you need to load the built-in overlay function:
```

```shell
#@ load("@ytt:overlay", "overlay")
```

In the `playground/basics/example-overlay-files` directory we have four files. 
Two are clipped Kubernetes YAML files and the other two have ytt functionality. 
We can see that both ops files are using the above load directive at the beginning. 
Then there is an overlay action to do a match so the operations are only performed on specific data. For these it’s checking the name of the resource e.g.
```
#@overlay/match by=overlay.subset({"metadata":{"name":"example-ingress1"}})
```

The other parts either are removing items or adding lines if they’re missing.
```yaml
#@overlay/remove
ingress.kubernetes.io/rewrite-target:

#@overlay/match missing_ok=True
nginx.ingress.kubernetes.io/enable-access-log: "true"
```

> Look at files and run in interactive playground:
> https://carvel.dev/ytt/#example:example-overlay-files 

Reviewer 2: ```suggestion
[Look at this example](https://carvel.dev/ytt/#example:example-overlay-files) and run it in the interactive playground
```

Author: Same thing here. The link is for the example, but it does not make it clear that the playground is on that link.

Or run locally:
```shell
ytt -f playground/basics/example-overlay-files
```

## Data Values and Schema
The data module, `#@ load("@ytt:data", "data")`, can be used to give users a configuration interface. It allows you to pass in files with a schema module, `#@data/values-schema`, or via command line. To see all of the options run `ytt -h` to see flags.

Note: If you are using an old version of ytt, the following ytt commands will fail.

> The k8s-add-global-label/add-label.yml file uses the annotation `​​#@ data.values.build_num` along with an overlay to check if it’s missing. If it’s missing it will be filled in with the line below.

Reviewer: "k8s-add-global-label/add-label.yml" should probably be formatted as in-line code (`k8s-add-global-label/add-label.yml`)

```
#@ load("@ytt:overlay", "overlay")
#@ load("@ytt:data", "data")

#@overlay/match by=overlay.all,expects="1+"
---
metadata:
  #@overlay/match missing_ok=True
  labels:
    #@overlay/match missing_ok=True
    build: #@ data.values.build_num
```

> The first run here will use build number from k8s-add-global-label/values.yml. There is a warning since the directory has a text file (it just contains the expected output). You could avoid this by passing in the three YAML files separately. Each one would need a `-f`. 

Reviewer 2: ```suggestion
The first run here will use build number from `k8s-add-global-label/values.yml`. There is a warning since the directory has a text file (it just contains the expected output). You could avoid this by passing in the three YAML files separately using the `-f` flag multiple times. 
```

```
#@data/values-schema
---
build_num: unknown
```

Run:
```shell
# uses defaults
ytt -f k8s-add-global-label
```

And the second run of it has the build number passed via a flag which will override it with “123”. 
```
# overrides the default
ytt -f k8s-add-global-label -v build_num=123

```

## Deploying to Kubernetes
ytt is purely a configuration tool. As you saw, ytt just outputs YAML files, but doesn’t apply or run them.
Let’s use these in conjunction with Kubernetes.
In order to deploy our applications to Kubernetes, we need to use tools like kubectl or [kapp (also part of the Carvel tool suite)](https://tanzu.vmware.com/developer/guides/kubernetes/kapp-gs/) with the output from ytt. 

> We have seen a bunch of different things we can do with ytt. The files in `playground/getting-started/example-overlay-on-templates` put together a lot of them. There is a README that explains everything that is happening here. You can also look at it on the playground:

Reviewer 2: ```suggestion
We have seen a bunch of different things we can do with ytt, and the files in `playground/getting-started/example-overlay-on-templates` put a lot of them together. There is a README that explains everything that is happening here. You can also look at it on the playground:
```
